[Xamarin.Android.Build.Tasks] first styleable resource id is always incorrect

dellis1972 opened the pull request:
Fixes #3049

This commit also reworks the id generation of the
`ManagedResourceParser` to try replicate the
ids which `aapt2` produces. As a result if the managed
parser generates the code from the `R.txt` file or
directly from the resources, the resulting designer.cs
file will be identical most of the time. There does seem
to be an odd ordering issue where the R.txt file does
not seem to order the items by name as the Id's are 
sometimes switched.

jonpryor commented:
The [`GenerateDesignerFileWithÜmläüts / Release`](https://jenkins.mono-project.com/job/xamarin-android-pr-pipeline-release/707/testReport/junit/Xamarin.Android.Build.Tests/ManagedResourceParserTests/GenerateDesignerFileWithÜmläüts___Release/) test failed.

jonpryor commented on src/Xamarin.Android.Build.Tasks/Utilities/ResourceIdentifier.cs:
```diff
@@ -0,0 +1,68 @@
+
+		private const string Identifier = IdentifierStartCharacter + "((" + IdentifierPartCharacter + ")+)*";
+
+		//https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#keywords
```
If we're going to "blacklist" identifiers, then we should also blacklist *Java* identifiers: https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-3.9

dellis1972 replied:
Not sure that will work. Its only the C# identifiers we need to prefix with a `@`.

jonathanpeppers replied:
Do we just need to make sure there is an `@` symbol for these?

For example, I can do:
```
public const int @class = 0;
```

dellis1972 replied:
@jonathanpeppers that is handled by https://github.com/xamarin/xamarin-android/pull/3025/files#diff-fe72c1d40099e8bd2cfa00d8eeccad39R63

jonpryor commented on src/Xamarin.Android.Build.Tasks/Utilities/ResourceIdentifier.cs:
```diff
@@ -0,0 +1,68 @@
+
+			var normalizedIdentifier = identifier.Normalize ();
+
+			string result = validIdentifier.Replace (normalizedIdentifier, "_");
```
I'm not at all sure I understand what this line *does* or why it does it. :-/

dellis1972 replied:
The validIdentifier is the negation of the actual regex we build up in the `Identifier` string. since we combine it with a `[^ ... ]`. This means we are saying, replace anything that does NOT match what is in the RegEx.

dellis1972 commented:
@jonathanpeppers I added the additional test case. 

```
<TextView android:id=""@+id/textview.withperiod"" />
<TextView android:id=""@+id/Føø-Bar"" />
```

These cover all of the cases you mentioned I believe. 

Also updated the system to be based on the C# spec for Identifiers.

jonpryor commented on src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Expected/GenerateDesignerFileExpected.cs:
```diff
@@ -27,8 +27,8 @@ static Resource()
 		public partial class Animator
 		{
 			
-			// aapt resource value: 0x7F070002
```
Why did our id values change?

dellis1972 replied:
Because we added two new fields. That changes the number of IDs we have and also the order in which they come.

dellis1972 replied:
Looking at the change, it seems that the order the files were processed in changed. I need to look at making sure we do this in a specific order to make sure its consistent.

dellis1972 replied:
I figured out a way to make the id's more consistent. They *should* match those being produced by `aapt` now.

dellis1972 replied:
You can confirm this by looking at the Ids in the Rtxt file contents https://github.com/xamarin/xamarin-android/pull/3025/files#diff-8649defe4caffc0b9f604d92c1ad8808R78. These ids were generated by aapt2 on the same resource structure. I also included a test to make sure we produce the same results from the Rtxt as well.

dellis1972 commented:
@jonpryor I need to update the commit message since we don't use the CodeDom anymore

jonathanpeppers commented on src/Xamarin.Android.Build.Tasks/Utilities/ResourceIdentifier.cs:
```diff
@@ -0,0 +1,68 @@
+			string result = validIdentifier.Replace (normalizedIdentifier, "_");
+
+			if (_keywords.Contains (result))
+				return $"@{result}";
```
Would it help if `_keywords` was a `HashSet<T>` and it used `Ordinal`? I think this `.Contains()` is from `System.Linq` and it probably iterates over the array.

dellis1972 replied:
Done

jonpryor commented:
[`GenerateDesignerFileWithÜmläüts / Release`](https://jenkins.mono-project.com/job/xamarin-android-pr-pipeline-release/733/testReport/Xamarin.Android.Build.Tests/ManagedResourceParserTests/GenerateDesignerFileWithÜmläüts___Release/) is failing:

```
  /Users/builder/jenkins/workspace/xamarin-android-pr-pipeline-release/xamarin-android/bin/TestRelease/temp/GenerateDesignerFileWithÜmläüts Some Space/Resource.designer.cs and /Users/builder/jenkins/workspace/xamarin-android-pr-pipeline-release/xamarin-android/bin/TestRelease/Expected/GenerateDesignerFileExpected.cs do not match.
  Expected: True
  But was:  False
```

dellis1972 commented:
I've found some issues with the parser. On more complex R.txt files the output is not identical. Especially for styleable stuff. For example take the following which is generated from an R.txt

```
public static int[] CoordinatorLayout_Layout = new int[] {
					16842931,
					2130903232,
					2130903233,
					2130903234,
					2130903237,
					2130903238,
					2130903239};
``` 

The first entry here has a weird value. That is because it equates to the `android:layout_gravity` resource.For our generator we can get that value from `platforms/android-X/data/res/values/public.xml`, and we will need to do that in order to generate an identical file.

There are also weird ordering issues when comparing the output. Take `borderWidth` and `borderlessButtonStyle`. in the R.txt then come in the order

```
// aapt resource value: 0x7F03003D
 public const int borderlessButtonStyle = 2130903101;

 // aapt resource value: 0x7F03003C
public const int borderWidth = 2130903100;
```
Noticed however that the VALUE of the items are out of order. Its like that sort the output only after generating the id's.

jonathanpeppers commented:
@dellis1972 can you rebase this one? Windows builds are fixed now.

dellis1972 commented:
@jonpryor This is good to go (needs a review). I think we might need this for 16-2 as well.

jonpryor commented on src/Xamarin.Android.Build.Tasks/Utilities/ResourceIdentifier.cs:
```diff
@@ -16,18 +16,18 @@ public class ResourceIdentifier {
-			DecimalDigitCharacter +
-			ConnectingCharacter +
-			CombiningCharacter +
+		private const string IdentifierPartCharacter = LetterCharacter + "," +
```
I believe we *don't* want to add `","` to this list or the previous lists:

```csharp
string FormattingCharacter = @"\p{Cf}";
string ConnectingCharacter = @"\p{Pc}";
string DecimalDigitCharacter = @"\p{Nd}";
string CombiningCharacter = @"\p{Mn},\p{Mc}";
string LetterCharacter = @"\p{Lu},\p{Ll},\p{Lt},\p{Lm},\p{Lo},\p{Nl}";
string IdentifierPartCharacter = LetterCharacter + "," +
	DecimalDigitCharacter + "," + ConnectingCharacter + "," +
	CombiningCharacter + "," + FormattingCharacter;
string IdentifierStartCharacter = "(" + LetterCharacter + ",_)";
string Identifier = IdentifierStartCharacter + "((" + IdentifierPartCharacter + ")+)*";
var r = new Regex ($"[^{Identifier}]", RegexOptions.Compiled);
r.Replace ("hello,world", "_");
// "hello,world"
```

Note that the `,` isn't replaced.

If we fixup *all* these declarations so that `,` is *not* included, then it replaces as I'd expect:

```csharp
string FormattingCharacter = @"\p{Cf}";
string ConnectingCharacter = @"\p{Pc}";
string DecimalDigitCharacter = @"\p{Nd}";
string CombiningCharacter = @"\p{Mn}\p{Mc}";
string LetterCharacter = @"\p{Lu}\p{Ll}\p{Lt}\p{Lm}\p{Lo}\p{Nl}";
string IdentifierPartCharacter = LetterCharacter +
	DecimalDigitCharacter + ConnectingCharacter +
	CombiningCharacter + FormattingCharacter;
string IdentifierStartCharacter = "(" + LetterCharacter + "_)";
string Identifier = IdentifierStartCharacter + "((" + IdentifierPartCharacter + ")+)*";
var r = new Regex ($"[^{Identifier}]", RegexOptions.Compiled);
r.Replace ("hello_world", "_");
```

dellis1972 replied:
Ah crap! the rebase broke stuff :( I'll fix it

jonpryor commented:
Suggested commit message:

```
Fixes: https://github.com/xamarin/xamarin-android/issues/3049

The `ManagedResourceParser` didn't support the use of `android:`-
referenced resource names.  Consequently, if such a resource is used
-- even implicitly! -- then the value that was inserted into
`Resource.designer.cs` is wrong, with a value of -1:

	<!-- styleable -->
	<resources>
	  <declare-styleable name="MvxImageView2">
	    <attr name="throttle" format="integer"/>
	    <attr name="imageUrl" format="string"/>
	    <attr name="imageViewTraceEnabled" format="boolean"/>
	    <attr name="throttleSvg" format="string"/>
	  </declare-styleable>
	</resources>

	// Resource.designer.cs
	partial class Styleable {
	  // aapt resource value: 0x1
	  public const int SvgImageView_android_adjustViewBounds = 1;
	  
	  // aapt resource value: 0x2
	  public const int SvgImageView_android_autoMirrored = 2;
	  
	  // aapt resource value: 0x7F010769
	  public const int SvgImageView_android_gravity = 2130773865;
	  // ...
	}

The fix is to add support for parsing `data/res/values/public.xml`
from an Android API level directory, allowing it to extract and use
the `/resources/public/@id` value for the desired `android:name`.
This allows `ManagedResourceParser` to generate correct Android
Resource id values, which results in the same id values as what
`aapt2` produces.

As a result if the managed parser generates the code from the `R.txt`
file or directly from the resources, the resulting
`Resource.designer.cs` file will be identical most of the time. There
does seem to be an odd ordering issue where the `R.txt` file does not
seem to order the items by name as the Id's are sometimes switched.
```
